examples/cosmos/ics20/ics20.qnt

```diff
@@ -146,90 +219,266 @@ module ics20Test {
       channels: channelCounterparties.get(chain),
-      channelEscrowAddresses: channelCounterparties.get(chain).keys().mapBy(_ => "escrow_account"),
-      outPackets: Set()
+      channelEscrowAddresses: channelCounterparties.get(chain).keys().mapBy(_ => ESCROW_ACCOUNT),
+      outPackets: Set(),
+      receivedButUnacknowledgedPackets: Set(),
+      inAcknowledgements: Set()
     })
   }
 
-  /// Receives a packet sent from `sourceChain` to `destChain`
-  action receivePacket(sourceChain: str, destChain: str): bool = {
+  /// Send a packet from `sourceChain` to `destChain`, indicating a transfer
+  /// of `amount`-many tokens of `denom` from `sender` to `receiver`.
+  action sendPacket(sourceChain: str, destChain: str,
+                    denom: DenomTrace, amount: UINT256,
+                    sender: Address, receiver: Address): bool = all {
+    // (1) Pre-condition:
+    // ICS 4: "a channel is a pipeline [...] between specific modules on **separate** blockchains"
+    sourceChain != destChain,
+    // there is a channel between `sourceChain` and `destChain`
+    CHANNEL_TOPOLOGY.keys().contains(sourceChain),
+    CHANNEL_TOPOLOGY.get(sourceChain).keys().contains(destChain),
+
+    // (2) Send the packet using `sendFungibleTokens`:
+    val sourceChainState = chainStates.get(sourceChain)
+    val newSourceChainState = sendFungibleTokens(sourceChainState, denom, amount,
+                                                 sender, receiver,
+                                                 "transfer", CHANNEL_TOPOLOGY.get(sourceChain).get(destChain),
+                                                 0, 0)
+
+    // (3) Quint state transition:
+    chainStates' = chainStates.set(sourceChain, newSourceChainState)
+  }
+
+  /// Receive a packet sent from `sourceChain` to `destChain`:
+  /// - Pick an arbitrary, previously unreceived packet sent from `sourceChain`
+  ///   to `destChain`.
+  /// - Call the `onRecvPacket` callback on it.
+  /// - Update the chain states to record the packet as received.
+  /// - Record (but not receive) the acknowledgement produced by `onRecvPacket`
+  ///   on `destChain`.
+  action receivePacket(sourceChain: str, destChain: str): bool =
     val sourceChainState = chainStates.get(sourceChain)
-    nondet packet = sourceChainState.outPackets.oneOf()
-    val result = onRecvPacket(chainStates.get(destChain), packet)
     all {
-      chainStates' = chainStates
-        .set(destChain, result._1)
-        .set(sourceChain, sourceChainState.with("outPackets", sourceChainState.outPackets.exclude(Set(packet))))
+      // (1) Pre-condition:
+      // ICS 4: "a channel is a pipeline [...] between specific modules on **separate** blockchains"
+      sourceChain != destChain,
+      // there is a channel between `sourceChain` and `destChain`
+      CHANNEL_TOPOLOGY.keys().contains(sourceChain),
+      CHANNEL_TOPOLOGY.get(sourceChain).keys().contains(destChain),
```

Comment:
This checks are already done by `isTravelingBetween` in line 300. Can we remove the redundancy or do you think it is important somehow?
```suggestion
```

Reply:
I feel like it belongs on the `isTravelingBetween` as a sanity check, not here. My interpretation is that if there is a packet traveling between A and B, then surely there is a connection between A and B. This check seems more like a sanity check on the compatibility of the packet with the topology, and since `isTravelingBetween` is the operator acting on the packet here, seems like sanity checks should be done there.

But this is not very relevant to me, feel free to keep the check here if you think its better.

Reply:
Then let's keep the redundance, since we both believe it's relevant at a different point :)